src/irmin-tezos-utils/parse.ml

```diff
@@ -0,0 +1,149 @@
+open Import
+open Files
+module Files = Make (Irmin_tezos.Conf) (Irmin_tezos.Schema)
+
+module Ring = struct
+  type t = { n : int; mutable i : int; a : int array }
+
+  let make n = { n; i = 0; a = Array.make n (-1) }
+
+  let add r x =
+    Array.set r.a r.i x;
+    r.i <- (r.i + 1) mod r.n
+
+  let get r i =
+    if i <= r.n then
+      let x = r.a.((r.i - i + r.n) mod r.n) in
+      if x <> -1 then Some x else None
+    else None
+end
+
+type ctx = { off : Int63.t; info : info }
+and info = { commit : int list; contents : int list; inode : int list }
+
+let info () = { commit = []; contents = []; inode = [] }
+
+type info_ring = { commit : Ring.t; contents : Ring.t; inode : Ring.t }
+
+let info_ring () =
+  { commit = Ring.make 1000; contents = Ring.make 1000; inode = Ring.make 1000 }
+
+let ctx_buffer = Bytes.create (4096 * 4096)
+
+let dump_ctx fd ctx =
+  let idx = ref 0 in
+  let f s =
+    String.iteri (fun i c -> Bytes.set ctx_buffer (!idx + i) c) s;
+    idx := !idx + String.length s
+  in
+  let flag =
+    List.length ctx.info.commit
+    + Int.shift_left (List.length ctx.info.contents) 2
+    + Int.shift_left (List.length ctx.info.inode) 4
+  in
+  Varint.encode_bin flag f;
+  List.iter (fun v -> Varint.encode_bin v f) ctx.info.commit;
+  List.iter (fun v -> Varint.encode_bin v f) ctx.info.contents;
+  List.iter (fun v -> Varint.encode_bin v f) ctx.info.inode;
```

Comment:
you could rely on repr and do `encode_bin ctx f` directly?

Reply:
I wanted to have a more precise and concise implementation as the lists are 3 elements max (hence my usage of only a few bits for the flag)
But I suppose we can yes

Reply:
maybe then write it (and encode it) as a triple?